### PR TITLE
[Docs] Translate embed/go/ref.md to zh. (#1092)

### LINKED ISSUE #1092: [Docs] Translate embed/go/ref.md to zh
### The MD file link you want to translate (must be in English) 
https://github.com/WasmEdge/WasmEdge/blob/master/docs/book/en/src/embed/go/ref.md

### Target language
https://github.com/WasmEdge/WasmEdge/blob/master/docs/book/zh/src/embed/go/ref.md

### PR DESCRIPTION
**Pull request by swartz-k**
fix #1092 

Signed-off-by: swartz-k <9215868@gmail.com>

### CODE REVIEW

**Comment by alabulei1** [comment]
@swartz-k  Thanks for your contribution! You really did a lot of work. I couldn't load the file now. Could you check out my comments firstly?

**Comment by swartz-k** [comment]
> @swartz-k Thanks for your contribution! You really did a lot of work. I couldn't load the file now. Could you check out my comments firstly?

Rebased lateset commit!

**Comment by alabulei1** [comment]
@swartz-k  I will check out tomorrow.

File: docs/book/zh/src/embed/go/ref.md [hunk 1]
```diff
@@ -1 +1,2155 @@
...
+
+```bash
+rustup default 1.50.0
+cd rust_bindgen_funcs
+rustwasmc build
+# 生成的 WASM 在 pkg/rust_bindgen_funcs_lib_bg.wasm
+```
+
+
+在 WasmEdge 中运行 WebAssembly 函数的 [Golang源码](https://github.com/second-state/WasmEdge-go-examples/blob/master/go_BindgenFuncs/bindgen_funcs.go) 如下。`ExecuteBindgen()` 函数调用 WebAssembly 函数并在 `wasm-bindgen` 支持下传递参数。
```
Comment (alabulei1) [inline docs/book/zh/src/embed/go/ref.md:162]: Golang  后面空一格。

File: docs/book/zh/src/embed/go/ref.md [hunk 2]
```diff
@@ -1 +1,2206 @@
...
+    const (
+        // 禁用尽可能多的优化。
+        CompilerOptLevel_O0 = CompilerOptimizationLevel(C.WasmEdge_CompilerOptimizationLevel_O0)
+        // 在不破坏可调试性的情况下快速优化。
+        CompilerOptLevel_O1 = CompilerOptimizationLevel(C.WasmEdge_CompilerOptimizationLevel_O1)
+        // 在不触发显着增量编译时间或代码大小增长的情况下尽可能地优化快速执行。
+        CompilerOptLevel_O2 = CompilerOptimizationLevel(C.WasmEdge_CompilerOptimizationLevel_O2)
+        // 尽可能优化快速执行。
+        CompilerOptLevel_O3 = CompilerOptimizationLevel(C.WasmEdge_CompilerOptimizationLevel_O3)
+        // 尽可能优化小代码大小，而不会触发显着的编译时间增量或执行时间减慢。
```
[the remark this answers is not in the document]
Comment (alabulei1) [reply docs/book/zh/src/embed/go/ref.md:720]: 这里没改哈

File: docs/book/zh/src/embed/go/ref.md [hunk 3]
```diff
@@ -1 +1,2155 @@
...
+        CompilerOptLevel_Os = CompilerOptimizationLevel(C.WasmEdge_CompilerOptimizationLevel_Os)
+        // 尽可能优化小代码大小。
```
Comment (alabulei1) [inline docs/book/zh/src/embed/go/ref.md:722]: 这里也没改哈

File: docs/book/zh/src/embed/go/ref.md [hunk 4]
```diff
@@ -1 +1,2155 @@
...
+    ```
+
+## WasmEdge VM
+
+在本部分中，我们将介绍 `wasmedge.VM` 对象的功能，并展示执行 WASM 功能的示例。
+
+### 带有 VM 对象的 WASM 执行示例
+
+下面是使用 WASM 运行斐波那契的示例。
+这个例子使用了 [fibonacci.wasm](../tools/wasmedge/examples/fibonacci.wasm)，对应的 WAT 文件在 [fibonacci.wat](../tools/wasmedge/examples/fibonacci.wat) .
```
Comment (alabulei1) [inline docs/book/zh/src/embed/go/ref.md:776]: . 应该是 。 不用空一格

File: docs/book/zh/src/embed/go/ref.md [hunk 5]
```diff
@@ -1 +1,2155 @@
...
+    procconf := conf.GetImportObject(wasmedge.WasmEdge_PROCESS)
+    // 初始化 WasmEdge_Process。
+    procconf.InitWasmEdgeProcess(/* ... 忽略 */)
+
+    conf.Release()
+    ```
+
+   也可以从 API 创建 WasmEdge_Process 导入对象。详细内容将在[host 函数](#Host-Functions）和 [主机模块注册](#Host-Module-Registrations) 中介绍。
+
+### 主机模块注册
```
Comment (alabulei1) [inline docs/book/zh/src/embed/go/ref.md:1070]: 主机还是不翻译？

File: docs/book/zh/src/embed/go/ref.md [hunk 6]
```diff
@@ -1 +1,2155 @@
...
+
+[host 函数](https://webassembly.github.io/spec/core/exec/runtime.html#syntax-hostfunc) 是 WebAssembly 之外的函数，并作为导入传递给 WASM 模块。 在 WasmEdge-go 中，host 函数被组合成宿主模块作为带有模块名称的 `ImportObject` 对象。详情请参考 [WasmEdge Runtime 中的主机函数](#Host-函数)。在本章中，我们展示了将 host 函数注册到 `VM` 对象中的示例。
```
Comment (alabulei1) [inline docs/book/zh/src/embed/go/ref.md:1072]: 主机改成 Host，还是不翻译？

File: docs/book/zh/src/embed/go/ref.md [hunk 7]
```diff
@@ -1 +1,2155 @@
...
+    // 从 store 对象中检索到的 `funcobj` 应该 __NOT__ 被释放。
+    // 从 `funcobj` 检索到的 `functype` 应该 __NOT__ 被释放。
+    ```
+
+   1. 表实例
+   
+       在 WasmEdge 中，你可以创建 `Table` 对象并将它们添加到 `ImportObject` 对象中，以便注册到 `VM` 或 `Store` 中。   
+       `Table` 对象提供 API 来控制表实例中的数据。
+
+       ```去
```
Comment (alabulei1) [inline docs/book/zh/src/embed/go/ref.md:1650]: 这里应该是 go 吧

File: docs/book/zh/src/embed/go/ref.md [hunk 8]
```diff
@@ -1 +1,2155 @@
...
+
+    meminst.Release()
+    ```
+
+3. 全局实例
+
+   在 WasmEdge 中，你可以创建 `Global` 对象并将它们添加到 `ImportObject` 对象中，以便注册到 `VM` 或 `Store` 中。    
+   `Global` 对象提供 API 来控制全局实例中的值。
+
+    ```去
```
Comment (alabulei1) [inline docs/book/zh/src/embed/go/ref.md:1732]: 这里应该是 go

File: docs/book/zh/src/embed/go/ref.md [hunk 9]
```diff
@@ -1 +1,2155 @@
...
+    // 如果值类型不匹配或
+    // 全局可变性是 `wasmedge.ValMut_Const`。
+    gotval := globinst.GetValue()
+    // `gotbal` 将是 `interface{}` 类型为 `uint64` 和
+    // 值为 888。
+
+    globinst.Release()
+    ```
+
+### 主机函数
```
Comment (alabulei1) [inline docs/book/zh/src/embed/go/ref.md:1759]: Host 不翻译？

File: docs/book/zh/src/embed/go/ref.md [hunk 10]
```diff
@@ -1 +1,2155 @@
...
+
+[主机函数](https://webassembly.github.io/spec/core/exec/runtime.html#syntax-hostfunc) 是 WebAssembly 之外的函数，并作为导入传递给 WASM 模块。
```
Comment (alabulei1) [inline docs/book/zh/src/embed/go/ref.md:1761]: Host 不翻译？

File: docs/book/zh/src/embed/go/ref.md [hunk 11]
```diff
@@ -1 +1,2155 @@
...
+1. 主机功能分配
+
+    你可以定义具有以下函数签名的 Go 函数作为宿主函数体：
+
+    ```go
+    type hostFunctionSignature func(
+        data interface{}, mem *Memory, params []interface{}) ([]interface{}, Result)
+    ```
+
+    添加 2 个 `i32` 值的 `add` 主机函数示例：
```
Comment (alabulei1) [inline docs/book/zh/src/embed/go/ref.md:1773]: Host 应该不翻译？

File: docs/book/zh/src/embed/go/ref.md [hunk 12]
```diff
@@ -1 +1,2155 @@
...
+
+    globinst.Release()
+    ```
+
+### 主机函数
+
+[主机函数](https://webassembly.github.io/spec/core/exec/runtime.html#syntax-hostfunc) 是 WebAssembly 之外的函数，并作为导入传递给 WASM 模块。   
+在 WasmEdge-go 中，你可以创建 `Function`、`Memory`、`Table` 和 `Global` 对象并将它们添加到 `ImportObject` 对象中，以便注册到 `VM` 或 `Store` 中。
+
+1. 主机功能分配
```
Comment (alabulei1) [inline docs/book/zh/src/embed/go/ref.md:1764]: Host 不翻译？

File: docs/book/zh/src/embed/go/ref.md [hunk 13]
```diff
@@ -1 +1,2155 @@
...
+    // 如果不需要外部数据，则为 `nil`。
+    // 如果你不需要成本计量，最后一个参数可以为0。
+    func_add := wasmedge.NewFunction(functype, host_add, nil, 0)
+
+    // 如果函数对象未添加到导入对象对象中，则应将其释放。
+    func_add.Release()
+    functype.Release()
+    ```
+
+2. 导入object对象
```
Comment (alabulei1) [inline docs/book/zh/src/embed/go/ref.md:1810]: object 应该去掉

File: docs/book/zh/src/embed/go/ref.md [hunk 14]
```diff
@@ -1 +1,2155 @@
...
+    // 创建一个函数实例并将其添加到导出名称为“add”的导入对象中。
+    functype := wasmedge.NewFunctionType(
+        []wasmedge.ValType{wasmedge.ValType_I32, wasmedge.ValType_I32},
+        []wasmedge.ValType{wasmedge.ValType_I32},
+    )
+    hostfunc := wasmedge.NewFunction(functype, host_add, nil, 0)
+    functype.Release()
+    impobj.AddFunction("add", hostfunc)
+
+    // 创建一个表实例并将其添加到导出名称为“table”的导入对象中。
```
Comment (alabulei1) [inline docs/book/zh/src/embed/go/ref.md:1840]: table 前后空一格

File: docs/book/zh/src/embed/go/ref.md [hunk 15]
```diff
@@ -1 +1,2155 @@
...
+            0x06, 0x61, 0x64, 0x64, 0x54, 0x77, 0x6F, 
+            /* 导出描述：函数 0 */
+            0x00, 0x01,
+            /* 代码部分 */
+            0x0A, 0x0A, 0x01, 
+            /* 代码主体 */
+            0x08, 0x00, 0x20, 0x00, 0x20, 0x01, 0x10, 0x00, 0x0B,
+        }
+
+        // 创建模块名称为“extern”的导入对象。
```
Comment (alabulei1) [inline docs/book/zh/src/embed/go/ref.md:1978]: “extern” 前后空一格

File: docs/book/zh/src/embed/go/ref.md [hunk 16]
```diff
@@ -1 +1,2155 @@
...
+        impobj := wasmedge.NewImportObject("extern")
+
+        // 创建一个函数实例并将其添加到导出名称为“func-add”的导入对象中。
```
Comment (alabulei1) [inline docs/book/zh/src/embed/go/ref.md:1981]: “func-add” 空一格

File: docs/book/zh/src/embed/go/ref.md [hunk 17]
```diff
@@ -1 +1,2155 @@
...
+        impobj.Release()
+        vm.Release()
+    }
+    ```
+
+   然后，你可以使用 WasmEdge Golang SDK 构建和运行 Golang 应用程序：
+
+    ```bash
+    $ 去获取 github.com/second-state/WasmEdge-go/wasmedge@v0.9.1
+    $去构建
```
Comment (alabulei1) [inline docs/book/zh/src/embed/go/ref.md:2009]: 去构建应该是 go build

File: docs/book/zh/src/embed/go/ref.md [hunk 18]
```diff
@@ -1 +1,2155 @@
...
+    ```go
+    package main
+
+    import (
+        "fmt"
+
+        "github.com/second-state/WasmEdge-go/wasmedge"
+    )
+
+    // 宿主函数体定义。
```
Comment (alabulei1) [inline docs/book/zh/src/embed/go/ref.md:2028]: host 不翻译

File: docs/book/zh/src/embed/go/ref.md [hunk 19]
```diff
@@ -1 +1,2155 @@
...
+            0x06, 0x61, 0x64, 0x64, 0x54, 0x77, 0x6F, 
+            /* 导出描述：函数 0 */
+            0x00, 0x01,
+            /* 代码部分 */
+            0x0A, 0x0A, 0x01, 
+            /* 代码主体 */
+            0x08, 0x00, 0x20, 0x00, 0x20, 0x01, 0x10, 0x00, 0x0B,
+        }
+
+        // 要设置到主机函数中的附加数据。
```
Comment (alabulei1) [inline docs/book/zh/src/embed/go/ref.md:2078]: host 不翻译

File: docs/book/zh/src/embed/go/ref.md [hunk 20]
```diff
@@ -1 +1,2155 @@
...
+        var 数据 int32 = 0
```
Comment (alabulei1) [inline docs/book/zh/src/embed/go/ref.md:2079]: 这里是代码，不用改翻译

数据 data

File: docs/book/zh/src/embed/go/ref.md [hunk 21]
```diff
@@ -1 +1,2155 @@
...
+
+        // 创建模块名称为“extern”的导入对象。
```
Comment (alabulei1) [inline docs/book/zh/src/embed/go/ref.md:2081]: “extern” 前后空一格

File: docs/book/zh/src/embed/go/ref.md [hunk 22]
```diff
@@ -1 +1,2155 @@
...
+        impobj := wasmedge.NewImportObject("extern")
+
+        // 创建一个函数实例并将其添加到导出名称为“func-add”的导入对象中。
```
Comment (alabulei1) [inline docs/book/zh/src/embed/go/ref.md:2084]: “func-add” 前后空一格

File: docs/book/zh/src/embed/go/ref.md [hunk 23]
```diff
@@ -1 +1,2155 @@
...
+    $ go get github.com/second-state/WasmEdge-go/wasmedge@v0.9.1
+    $ go build
+    $ ./wasmedge_test
+    Get the result: 6912
+    Data value: 6912
+    ```
+
+## WasmEdge AOT 编译器
+
+在本部分中，我们将介绍 WasmEdge AOT 编译器和 Go 中的选项。
```
Comment (alabulei1) [inline docs/book/zh/src/embed/go/ref.md:2119]: 这一行和下面的两行不应该断行，是一行。

File: docs/book/zh/src/embed/go/ref.md [hunk 24]
```diff
@@ -1 +1,2155 @@
...
+
+const (
+    // 原生动态库格式。
+    CompilerOutputFormat_Native = CompilerOutputFormat(C.WasmEdge_CompilerOutputFormat_Native)
+    // WebAssembly 与自定义部分中的 AOT 编译代码。
+    CompilerOutputFormat_Wasm = CompilerOutputFormat(C.WasmEdge_CompilerOutputFormat_Wasm)
+)
+```
+
+详情请参考 [AOT 编译器选项配置](#Configurations)。
```
Comment (alabulei1) [inline docs/book/zh/src/embed/go/ref.md:2155]: [AOT 编译器选项配置](#Configurations) 如果括号里的英文修改了，那括号里应该是中文。

**Review by alabulei1 — COMMENTED** [review]
@swartz-k Thanks for your contribution. Please check out my comments.

**Comment by alabulei1** [comment]
LGTM.

@swartz-k  thanks for your contribution.

@hydai  Could you help merge this PR?

**Review by hydai — CHANGES_REQUESTED** [review]
There are some problems from the previous feedback are not fixed. Please check the unresolved parts and fix them.

**Comment by swartz-k** [comment]
> There are some problems from the previous feedback are not fixed. Please check the unresolved parts and fix them.

I had fix. But cannot find solved button. Please check it.

File: docs/book/zh/src/embed/go/ref.md [hunk 25]
```diff
@@ -1 +1,2206 @@
...
+    // 假设 `conf` 是 `*wasmedge.Configure` 对象。
+
+    // 创建统计对象。如果统计数据，则无需此步骤
+    // 不需要。
+    stat := wasmedge.NewStatistics()
+    // 创建执行器对象。
+    // 使用默认配置创建执行器，没有统计信息，
+    // 你可以使用 `wasmedge.NewExecutor()` 代替。
+    executor := wasmedge.NewExecutorWithConfigAndStatistics(conf, stat)
+    // 创建存储对象。store 是 WASM 运行时结构的核心。
```
[the remark this answers is not in the document]
Comment (hydai) [reply docs/book/zh/src/embed/go/ref.md:1524]: 這個沒處理

File: docs/book/zh/src/embed/go/ref.md [hunk 26]
```diff
@@ -1 +1,2206 @@
...
+   在实例化 WASM 模块之前，请检查 [import section](https://webassembly.github.io/spec/core/syntax/modules.html#syntax-import) 以确保将导入注册到 `Store` 对象中.
+
+    ```go
+    // ...
+    // 假设 `ast` 是加载器的输出 `*wasmedge.AST` 对象
+    // 并且通过了验证。
+    // 假设 `conf` 是 `*wasmedge.Configure` 对象。
+
+    // 创建统计对象。如果统计数据，则无需此步骤
+    // 不需要。
```
[the remark this answers is not in the document]
Comment (hydai) [reply docs/book/zh/src/embed/go/ref.md:1518]: 這個沒處理

File: docs/book/zh/src/embed/go/ref.md [hunk 27]
```diff
@@ -1 +1,2206 @@
...
+2. 实例化模块
+
+   WASM 或已编译的 WASM 模块应在函数调用之前实例化。   
+   请注意，开发人员只能将一个模块实例化到 `Store` 对象中，在这种情况下，旧的实例化模块将被清除。    
+   在实例化 WASM 模块之前，请检查 [import section](https://webassembly.github.io/spec/core/syntax/modules.html#syntax-import) 以确保将导入注册到 `Store` 对象中.
+
+    ```go
+    // ...
+    // 假设 `ast` 是加载器的输出 `*wasmedge.AST` 对象
+    // 并且通过了验证。
```
[the remark this answers is not in the document]
Comment (hydai) [reply docs/book/zh/src/embed/go/ref.md:1514]: 這個沒處理

File: docs/book/zh/src/embed/go/ref.md [hunk 28]
```diff
@@ -1 +1,2206 @@
...
+            return
+        }
+        // "fibonacci.wasm" 中的函数 "fib" 与模块一起导出
+        // 名称“mod”。与主机功能一样，其他模块可以导入
+        // 函数 `"mod" "fib"`.
+
+        // 在注册模块中执行 WASM 函数。
+        // 与函数的执行不同，注册的函数可以
+        // 在没有 `(*VM).Instantiate` 的情况下调用，因为 WASM 模块是
+        // 注册时实例化。
```
[the remark this answers is not in the document]
Comment (hydai) [reply docs/book/zh/src/embed/go/ref.md:1152]: 這個沒處理

File: docs/book/zh/src/embed/go/ref.md [hunk 29]
```diff
@@ -1 +1,2206 @@
...
+        "github.com/second-state/WasmEdge-go/wasmedge"
+    )
+
+    func main() {
+        // 创建虚拟机。
+        vm := wasmedge.NewVM()
+
+       var res []interface{}
+        var err error
+        // 使用模块名称“mod”将 WASM 模块从文件注册到 VM。
```
[the remark this answers is not in the document]
Comment (hydai) [reply docs/book/zh/src/embed/go/ref.md:1140]: 這個沒處理

File: docs/book/zh/src/embed/go/ref.md [hunk 30]
```diff
@@ -1 +1,2152 @@
...
+    ```
+
+3. 最大内存页数
+
+   你可以通过这个配置来限制内存实例的页面大小。当在 WASM 执行中增加内存实例的页面大小并超过限制大小时，会失败。此配置仅在 `Executor` 和 `VM` 对象中有效。
+
+    ```go
+    conf := wasmedge.NewConfigure()
+
+	  pagesize := conf.GetMaxMemoryPage()
```
Comment (hydai) [inline docs/book/zh/src/embed/go/ref.md:696]: 縮排錯了

File: docs/book/zh/src/embed/go/ref.md [hunk 31]
```diff
@@ -1 +1,2206 @@
...
+    ```
+
+### 配置
+
+配置对象 `wasmedge.Configure` 管理 `Loader`、`Validator`、`Executor`、`VM` 和 `Compiler` 的配置。
+开发者可以调整 proposals 配置、VM 主机预注册（比如 `WASI`）和 AOT 编译器选项的设置，然后应用 `Configure` 对象来创建其他运行时对象。
+
+1. 配置
+
+   WasmEdge 支持打开或关闭 WebAssembly proposals。
```
[the remark this answers is not in the document]
Comment (hydai) [reply docs/book/zh/src/embed/go/ref.md:626]: 這個沒處理

**Review by hydai — CHANGES_REQUESTED** [review]

**Comment by swartz-k** [comment]
[docs/book/zh/src/embed/go/ref.md](https://github.com/WasmEdge/WasmEdge/pull/1198/files/5493f5ea4b39371c2e6e40908c4857da15df9a96#diff-64955e120ccd4eed3eb4a094695acd65bbbd8a3211554f1eabd404aaf56e9a24)

```
1. 配置

   WasmEdge 支持打开或关闭 WebAssembly proposals。

```

Sorry. I don't get how to deal with it.

**Comment by hydai** [comment]
截圖 2022-03-08 at 16 06 41@2x
Looks like the review system is not shown the details?

**Comment by swartz-k** [comment]
> 截圖 2022-03-08 at 16 06 41@2x
> 
> Looks like the review system is not shown the details?

Right. I fix them. Thanks~

**Comment by alabulei1** [comment]
@hydai I think we could merge this PR now. All the comments have been solved.

**Comment by hydai** [comment]
> @hydai I think we could merge this PR now. All the comments have been solved.

Got it. Sorry for missing this PR.

**Review by hydai — APPROVED** [review]

**Pull request merged**